adding collection editor and object selector editor

JuditRose opened the pull request:
This code is copied over from VS and Winforms Framework repos. So if something doesn't make sense here in core let me know and I am happy to change it.

Fixes #177 
contributes to #497 by removing SR files and ensuring they are autogenerated

zsd4yr commented on src/System.Windows.Forms/src/System/Windows/Forms/RTLAwareMessageBox.cs:
```diff
@@ -15,7 +15,7 @@ namespace System.Windows.Forms {
     ///    </para>
     /// </devdoc>
-    internal sealed class RTLAwareMessageBox {
+    public sealed class RTLAwareMessageBox {
```
Can you explain why this had to be / why you wanted to make this public?

JuditRose replied:
Because the editors I just added needed these 2 classes, but these classes are in System.Windows.Forms and the editors are in System.Windows.Forms.Design.Editors... I couldn't figure out a better way to make them visible there. But I am open for discussion.

JuditRose replied:
There is also a System.SR class in System.Windows.Forms, but that one doesn't contain most of the stuff I needed, so I believe it was worth to create a new class (with a same name) only for the design editors. 
But RTLAwareMessageBox  and ClientUtils classes contain the methods that the editors needed, so I didn't want to duplicate.

Tanya-Solyanik replied:
Alternative would be have a common utility assembly with InternalsViibleTo for System.Windows.Forms and .Editors. I think we have more classes that would logically belong to that assembly, see content of the common folder at the top level as well as https://github.com/dotnet/winforms/tree/master/src/System.Windows.Forms.Design.Editors/src/Misc and https://github.com/dotnet/winforms/tree/master/src/System.Windows.Forms/src/misc

We have to document public classes, and these two classes are very specialized to be useful for customers, I think

zsd4yr commented on src/System.Windows.Forms/src/misc/ClientUtils.cs:
```diff
@@ -23,7 +23,7 @@ namespace System.Windows.Forms
 
     // Miscellaneous utilities
-    static internal class ClientUtils {
+    static public class ClientUtils {
```
Same question as other

JuditRose replied:
Fixed both

JuditRose commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/ObjectSelectorEditor.cs:
```diff
@@ -0,0 +1,320 @@
+
+namespace System.Windows.Forms.Design.Editors.System.ComponentModel.Design
+{
+    [SuppressMessage("Microsoft.Design", "CA1012:AbstractTypesShouldNotHaveConstructors")]
```
I wasn't sure which SuppressMessages are safe to remove, so I kept them, but let me know if I can remove any of them

Tanya-Solyanik replied:
We probably want to keep FXCop suppression for the future, when we enable FxCop in this repo

JuditRose replied:
sure, I'll keep them then

JuditRose commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/ObjectSelectorEditor.cs:
```diff
@@ -0,0 +1,320 @@
+        /// Suppressed because although the type implements ISerializable --its on the base class and this class
+        /// is not modifying the stream to include its local information.  Therefore, we should not publicly advertise this as
+        /// Serializable unless explicitly required.
+        [SuppressMessage("Microsoft.Usage", "CA2237:MarkISerializableTypesWithSerializable")]
```
Probably this doesn't even make sense any more and just hasn't been cleaned yet?

Tanya-Solyanik replied:
It is still applicable, TreeNode implements ISerializable

JuditRose replied:
I'll keep this one then as it is

zsd4yr commented:
A search of our repo indicates that we use the SuppressMessage attribute on the following Microsoft.* binaries:
* Performance
* Reliability
* Design
* Usage
* Portability
* Interopability
* Reliability
* Security
* Globalization
* Naming

JuditRose commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/ExpandCollapseState.cs:
```diff
@@ -0,0 +1,18 @@
+
+using System.Runtime.InteropServices;
+
+namespace System.Windows.Forms.Design.Editors.System.ComponentModel.Design
```
I don’t really like this namespace, but this is the folder that these editors seemed to fit in. If you guys have any better idea how to define these namespaces, please let me know

Tanya-Solyanik replied:
We don't have to preserve folder structure of the "desktop" enlistment.

JuditRose commented:
> A search of our repo indicates that we use the SuppressMessage attribute on the following Microsoft.* binaries:
> 
> * Performance
> * Reliability
> * Design
> * Usage
> * Portability
> * Interopability
> * Reliability
> * Security
> * Globalization
> * Naming

Does that mean I can remove all the rest?

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/SR.cs:
```diff
@@ -0,0 +1,34 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+///------------------------------------------------------------------------------
```
please remove the extra comment

Tanya-Solyanik commented:
BTW, @bryukov is making changes in one of the editors (StringColelction?) in the VisualStudio. Dmitry, when you port the fix you are working on to Core, this is where it would go

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/ExpandCollapseState.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    [ComVisible(true)]
+    [Guid("76d12d7e-b227-4417-9ce2-42642ffa896a")]
+    public enum ExpandCollapseState
```
can this be internal?

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/ObjectSelectorEditor.cs:
```diff
@@ -0,0 +1,320 @@
+        }
+        
+        [SuppressMessage("Microsoft.Security", "CA2123:OverrideLinkDemandsShouldBeIdenticalToBase")] // everything in this assembly is full trust.
+        public override object EditValue(ITypeDescriptorContext context, IServiceProvider provider, object value)
```
comment can be removed

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/SR.cs:
```diff
@@ -0,0 +1,34 @@
+// See the LICENSE file in the project root for more information.
+
+///------------------------------------------------------------------------------
+
```
this file can be moved to the common folder, and the sr.cs in winforms can be deleted, if the winforms project references the common copy instead. But you would need to change the namespace to system.windows.forms, so that both project can be compiled.

JuditRose replied:
We talked about this, and we are keeping it out from common for now.

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2785 @@
+                            cs.ComponentChanging += new ComponentChangingEventHandler(OnComponentChanging);
+                        }
+
+                        //if (localCollectionForm.ShowEditorDialog(edSvc) == DialogResult.OK)
```
If this code was commented out in the "original", feel free to delete it.

JuditRose replied:
Talked to Zach, this few lines wasn't commented out originally, I just wasn't sure if it's needed, he said we need it, so I kept them

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2785 @@
+                {
+                    _currentContext = context;
+
+                    // Always create a new CollectionForm.  We used to do reuse the form in V1 and Everett
```
commented out code can be removed

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2785 @@
+                            _listbox.Items.Add(new ListItem(_editor, items[i]));
+                            _originalItems.Add(items[i]);
+                        }
+                        if (_listbox.Items.Count > 0)
```
extra spaces

Tanya-Solyanik replied:
Extra spaces

Tanya-Solyanik replied:
these comments can be removed from all properties in this class

zsd4yr commented:
> Does that mean I can remove all the rest? That would be great!

What are the ones that are in this PR that are not in that list? I do not necessarily think it means that.

JuditRose commented:
> > Does that mean I can remove all the rest? That would be great!
> 
> What are the ones that are in this PR that are not in that list? I do not necessarily think it means that.

Microsoft.Usage is the only one in this PR, but later on it would be useful to know which ones to keep or safe remove

JuditRose commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2785 @@
+            }
+        }
+
+        internal class SplitButton : Button
```
@Tanya-Solyanik Can you please help me to answer this question?

JuditRose commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2785 @@
+
+            foreach (object o in items)
+            {
+                if (o is IComponent comp && comp.Site == null)
```
Hi @RussKie ,
Thanks for the review. As I mentioned in the description this code is ported (from winforms framework version) not newly written. These components will be used by porting applications from dotnet framework to core. So I'd like to keep my own changes (fixes, cleanup) minimal, so no accidental functionality changes happen. 
I changed a few things according to your comments. For the ones I haven't changed: If you believe that functionality doesn't break but it's worth to change it because of better performance or readability, you are very welcome to open a PR after this one has been merged and the team will take a look.

RussKie replied:
Sure, no worries, fully understand.

JuditRose commented:
Hi @RussKie ,
Thanks for the review. As I mentioned in the description this code is ported (from winforms framework version) not newly written. These components will be used by porting applications from dotnet framework to core. So I'd like to keep my own changes (fixes, cleanup) minimal, so no accidental functionality changes happen.
I changed a few things according to your comments. For the ones I haven't fixed: If you believe that functionality doesn't break but it's worth to change it because of better performance or readability, you are very welcome to open a PR after this one has been merged and the team will take a look.

zsd4yr commented:
> > > Does that mean I can remove all the rest? That would be great!
> > 
> > 
> > What are the ones that are in this PR that are not in that list? I do not necessarily think it means that.
> 
> Microsoft.Usage is the only one in this PR, but later on it would be useful to know which ones to keep or safe to remove

Would you mind creating an issue for this / putting this topic on our tactics schedule for next week? I want to make sure we understand the full scope of these

JuditRose commented:
> Looks good. One thing that is missing is that public and protected types and members are not added to the reference assembly -
> 
> [winforms/src/System.Windows.Forms.Design.Editors/ref/System.Windows.Forms.Design.Editors.cs](https://github.com/dotnet/winforms/blob/66054e81885b65027a5bd0f013d3917bd1b735c8/src/System.Windows.Forms.Design.Editors/ref/System.Windows.Forms.Design.Editors.cs#L67)
> Line 67 in [66054e8](/dotnet/winforms/commit/66054e81885b65027a5bd0f013d3917bd1b735c8)
>  namespace System.Windows.Forms.Design 
> 
> and the same types should be removed from the app compat contract - https://github.com/dotnet/winforms/blob/master/src/System.Design/src/ApiCompatBaseline.netcoreapp.txt
> This could be generated by running a build target, please sync up with @dreddy-work.

I updated the ref cs file to contain the new component. I just noticed that you mentioned namespace S.W.F.Design, I added them under S.W.F.Design.Editors, as they are defined there. Let me know if I need to change it.
Also I talked to Devendar and he says that we are not doing the generation any more, so I updated the txt manually and removed the new component from there.

JuditRose commented:
I there a way that I can test these components before merging?

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/ref/System.Windows.Forms.Design.Editors.cs:
```diff
@@ -147,3 +151,82 @@ public partial class ShortcutKeysEditor : System.Drawing.Design.UITypeEditor
+        public class SelectorNode : TreeNode
+        {
+            public SelectorNode(string label, object value) : base(label) { }
+
```
extra space

Tanya-Solyanik commented on src/Common/src/RTLAwareMessageBox.cs:
```diff
@@ -1,10 +1,15 @@
 
 namespace System.Windows.Forms {
-    using System;
+    #if WINFORMS_NAMESPACE
```
I think this can be simplified:

    #if WINFORMS_NAMESPACE
        using System;
    #else
        using System.Windows.Forms.Design.Editors.System;
    #endif

JuditRose replied:
I replaced them, but it was breaking my local build, so I changed it back

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -290,21 +290,20 @@ protected virtual void DestroyInstance(object instance)
                     // child modal dialog -launching in System Aware mode 
                     CollectionForm localCollectionForm = DpiHelper.CreateInstanceInSystemAwareContext(() => CreateCollectionForm());
                     ITypeDescriptorContext lastContext = _currentContext;
-                    //localCollectionForm.EditValue = value;
```
Is this code enabled in the VS repo?

Tanya-Solyanik commented:
> I just noticed that you mentioned namespace S.W.F.Design, I added them under S.W.F.Design.Editors, as they are defined there. Let me know if I need to change it.

You are right, I had mistyped the namespace name 

> Is there a way that I can test these components before merging?

Yes, create a new CollectionEditor from winforms core application and call .ShowDialog on it.

dreddy-work commented on src/System.Windows.Forms.Design.Editors/src/System/SR.cs:
```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
I am assuming this file was auto generated?

JuditRose replied:
No, I created it. is there a way to generate it?

Tanya-Solyanik replied:
some versin of this file is generated in netfxdev1, you can find it in some intermediate directory in binaries folder

dreddy-work commented on src/System.Windows.Forms/src/System.Windows.Forms.csproj:
```diff
@@ -22,6 +22,9 @@
 
   <ItemGroup>
     <Compile Include="..\..\Common\src\ApiHelper.cs" Link="Common\ApiHelper.cs" />
+    <Compile Include="..\..\Common\src\ClientUtils.cs" Link="Common\ClientUtils.cs" />
```
Are these references used in this dll?

JuditRose replied:
yes

dreddy-work commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2666 @@
+
+namespace System.Windows.Forms.Design.Editors
+{
+    /// <devdoc>
```
"devdoc" is replaced by ///summary tag in core.

JuditRose replied:
replaced all occurrences, thanks for noticing

dreddy-work commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/CollectionEditor.cs:
```diff
@@ -0,0 +1,2666 @@
+                        }
+                        else if (_listbox.Items.Count > 0)
+                        {
+                            _listbox.Selected
```
Can you please make sure some basic description for all public/protected methods?

JuditRose replied:
I copied over all the descriptions that were present previously. I can definitely add something basic description, but based on my knowledge it would be something like 
```
///<summary>
/// Get the selected item
///</summary>
public SelectedItem GetSelected (...) {...}
```
But honestly it doesn't seem to be useful to me, so I'd rather leave it as it is until we can provide a meaningful description. 
Let me know if it's still preferred to have all the docs, even if their usefulness is questionable...

dreddy-work commented on src/System.Windows.Forms.Design.Editors/ref/System.Windows.Forms.Design.Editors.cs:
```diff
@@ -1,11 +1,15 @@
 // Changes to this file must follow the http://aka.ms/api-review process.
 // ------------------------------------------------------------------------------
 
+using System.Collections;
```
Can you please add some basic unit-tests as we progress with porting of this new source?

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/Resources/SR.resx:
```diff
@@ -1,5 +1,64 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
+  <!-- 
```
.Net Core team had requested that this commented out header is removed, Devendar might have details as to why.

danmosemsft replied:
It is just clutter, but if an edit in the VS resource editor forces or back then (unless and until the editor is fixed) it is probably best to give in.

Tanya-Solyanik replied:
@danmosemsft - unfortunately resxresourcewriter that writes this header is in the classic FX, and I don't think this change makes the servicing bar, so as long as VS runs on the classic FX, this header will be re-generated by the resource editor

JuditRose replied:
This change has been generated by VS, do we really want to remove it manually every time after we change this file?

zsd4yr replied:
Perhaps we can gitignore this change or do something fancy in gitattributes?

RussKie replied:
Gitignore will make git to ignore a file completely, and should you need to commit changes to the file you will need to add the file explicitly. There are few other ways to ignore files (like `skipworktree` and `assume unchanged`), but I don't see how they help in this case either.

Gitattributes inform git on how to handle a file (e.g. text vs binary), but it can't help with ignoring content.

zsd4yr replied:
> Gitignore will make git to ignore a file completely, and should you need to commit changes to the file you will need to add the file explicitly. There are few other ways to ignore files (like `skipworktree` and `assume unchanged`), but I don't see how they help in this case either.
> 
> Gitattributes inform git on how to handle a file (e.g. text vs binary), but it can't help with ignoring content.

Thanks for the education hmmm.... shame there's no obvious solution. The likelyhood of us touching this in classic FX is nearly zero

RussKie replied:
You could probably devise some arcane scripts or hooks, but would it worth it? 

A shameless plug, GitExtensions allows you to see all kinds of files:
![image](https://user-images.githubusercontent.com/4403806/53222862-686ce380-36c3-11e9-8800-c4f7e46efd42.png)

zsd4yr commented on src/System.Windows.Forms.Design.Editors/tests/UnitTests/ObjectSelectorEditorTests.cs:
```diff
@@ -0,0 +1,75 @@
+
+        private ObjectSelectorEditor GetNewObjectSelectorEditor(bool subObjectSelector = false)
+        {
+            return subObjectSelector ? new TestObjectSelectorEditor(subObjectSelector) : new TestObjectSelectorEditor();
```
You should use the Moq framework instead. This is not a recommend method, and it adds unnecessary bloat to the file

For an example, please see https://github.com/dotnet/winforms/blob/f25dc3d526b121ab9adbff3b5779c856a07afa78/src/System.Windows.Forms/tests/UnitTests/ContextMenuStrip.cs#L25

JuditRose replied:
But this is the class I am testing, I don't really want to mock that

zsd4yr replied:
@JuditRose you should not create TestObjectSelectorEditor which extends ObjectSelectorEditor. Instead, you should mock ObjectSelectorEditor. Creating a child class just for the purposes of testing is not recommended.

Tanya-Solyanik replied:
@JuditRose  - were you able to test your editor in an application with propertygrid?
@zsd4yr  - a derived class seems to be the right choice in this case, see https://github.com/dotnet/winforms/tree/master/src/System.Windows.Forms.Design.Editors/tests/UnitTests/Mocks, these is not a mock either, rather a derived class. Judit, consider moving your derived class to the same folder.

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/tests/UnitTests/ObjectSelectorEditorTests.cs:
```diff
@@ -0,0 +1,75 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
BTW, some files in the unit tests project have suffix `Tests` and some don't . What is the naming convention here? I think we have more files with no suffix, would be good to keep this consistent.

JuditRose replied:
I have seen files with Tests prefix and without it too. I'd prefer having Test or Tests in the name, instead of having the same name for both of them, because from the file name I can see which one is the implementation and which one is the test. Also search for a file or class is cleaner. But that is my personal preference. I can change it if the team prefers the other way.

RussKie replied:
I've been involved in many projects and traditionally test projects have "Tests" suffix in project names and in file names. 
This convention helps in reasoning about the codebase.

My 2c.

zsd4yr replied:
@RussKie I believe that is already the case https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms.Design.Editors/tests/UnitTests/System.Windows.Forms.Design.Editors.Tests.csproj

Tanya-Solyanik replied:
CoreFX goes with no suffixes: https://github.com/dotnet/corefx/tree/master/src/System.Reflection.Emit/tests/ConstructorBuilder, we copied our style from them initially, this convention shortens the path and these files are located in `tests` folder, so readability does not suffer much. However I primarily care about consistent naming convention in this repo, and is we go with `no suffix` convention, we have less files to rename.

RussKie replied:
Would you like me to send a PR with renames?

Tanya-Solyanik replied:
Sure :)

zsd4yr commented:
@Tanya-Solyanik @dreddy-work What else is left to be done here? Judit is gone for a bit but I am pretty sure we want this in :)

Tanya-Solyanik commented:
> @Tanya-Solyanik @dreddy-work What else is left to be done here? Judit is gone for a bit but I am pretty sure we want this in :)

My only queation is if Judit had a chanse to test a "live" scenario, that uses editor in a property gris

Tanya-Solyanik commented:
@dreddy-work  - could you please review this PR, I removed my block but I'm not sure if facades and refs are done correctly

zsd4yr commented:
@dreddy-work is there anything else to be done here?

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/ref/System.Windows.Forms.Design.Editors.cs:
```diff
@@ -147,3 +151,81 @@ public partial class ShortcutKeysEditor : System.Drawing.Design.UITypeEditor
+            public Selector(ObjectSelectorEditor editor) { }
+            public SelectorNode AddNode(string label, object value, SelectorNode parent) { throw null; }
+            public void Clear() { }
+            protected void OnAfterSelect(object sender, System.Windows.Forms.TreeViewEventArgs e) { }
```
Does `using System,Windows.Forms;` on the top of this file not work? Why did you have to add namespace names?

zsd4yr replied:
It may work. I did not try.

I noticed this is the pattern used elsewhere in this file so I followed it; besides, these should be autogenerated soon

Tanya-Solyanik commented on src/System.Windows.Forms.Design.Editors/src/Resources/SR.resx:
```diff
@@ -1,5 +1,64 @@
+    value   : The object must be serialized into a byte array 
+            : using a System.ComponentModel.TypeConverter
+            : and then encoded with base64 encoding.
+    -->
```
CoreFx team had requested that these headers are removed.

RussKie replied:
This discussion has already taken place - https://github.com/dotnet/winforms/pull/439#discussion_r258308346

zsd4yr replied:
Can this be done? I thought we determined this was autogenerated?

zsd4yr replied:
Done

zsd4yr replied:
Should I have a CollectionEditor.resx file in that same directory? That feels wrong

@dreddy-work

zsd4yr replied:
spoke offline. This is fine. MaskDesignerDialog, TreeNodeCollectionEditor, ToolStripItemCollectionEditor, StyleCollectionEditor, ShortcutKeyEditor, StringCollectionEditor, FormatControl, linkareaeditor, DataGridViewAutoFormatDialog, DataGridViewColumnCollectionDialog, DataGridViewCellStyleBuilderDialog, DataGridAutoFormatDialog, BorderSidesEditor, BindingFormattingDialog, CollectionEditor, and BinaryEditor are the only standalone forms / dialogues in the classic framework which have their own .resx files

dreddy-work commented on src/System.Windows.Forms.Design.Editors/src/Resources/CollectionEditor.resx:
```diff
@@ -0,0 +1,556 @@
+    <value>CollectionEditor</value>
+  </data>
+  <data name="&gt;&gt;$this.Type" xml:space="preserve">
+    <value>System.Windows.Forms.Form, System.Windows.Forms, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
```
Make sure this land in the localization.

dreddy-work replied:
Looks like it is already doing.

dreddy-work commented on src/System.Windows.Forms.Design.Editors/src/Resources/CollectionEditor.resx:
```diff
@@ -0,0 +1,556 @@
+  <data name="propertyBrowser.TabIndex" type="System.Int32, mscorlib">
+    <value>6</value>
+  </data>
+  <data name="&gt;&gt;propertyBrowser.Name" xml:space="preserve">
```
We need to check why this does not need to generate sources? Don't we have any constants defined here?

zsd4yr replied:
done

dreddy-work commented on src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/ObjectSelectorEditor.cs:
```diff
@@ -0,0 +1,320 @@
+                return false;
+            }
+            
+            public void Start(IWindowsFormsEditorService edSvc, object value)
```
Can we make sure we have some comment description for all public methods?

zsd4yr replied:
done

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/System.Windows.Forms.IntegrationTests/TestHelpers.cs:
```diff
@@ -208,7 +208,7 @@ public static bool PressTabsOnProcess(Process process, uint times)
         /// <seealso cref="System.Windows.Forms.SendKeys.SendWait(string)"/>
         /// <seealso cref="System.Threading.Thread.Sleep(int)"/>
-        internal static bool PressOnProcess(Process process, string keys)
+        public static bool PressOnProcess(Process process, string keys)
```
You are using this method in the same assembly it is declared, so it does not have to be public, no?

zsd4yr replied:
addressed

zsd4yr commented:
![image](https://user-images.githubusercontent.com/10677481/53599819-f54cec80-3b5c-11e9-9a1e-8597c81d72d5.png)
